### PR TITLE
refac: adding missing error wrapping.

### LINKED ISSUE #1059: Review all fmt.Errorf to use %w to allow internal error to be wrapped
`%w` allows the internal error to be `wrapped` by the external error, allowing methods like `errors.Is`, `errors.As` and `errors.Unwrap` to work properly, making it easier to check for specific errors on different layers of the code.

example: 
```go
package main

import (
	"errors"
	"fmt"
)

func main() {
	errLvl1 := fmt.Errorf("error 1")
	errLvl2 := fmt.Errorf("error 2, because of: %w", errLvl1)
	errLvl3 := fmt.Errorf("error 3, because of: %w", errLvl2)

	fmt.Println(errors.Is(errLvl1, errLvl1)) // true
	fmt.Println(errors.Is(errLvl1, errLvl2)) // false
	fmt.Println(errors.Is(errLvl1, errLvl3)) // false
	fmt.Println()
	fmt.Println(errors.Is(errLvl2, errLvl1)) // true
	fmt.Println(errors.Is(errLvl2, errLvl2)) // true
	fmt.Println(errors.Is(errLvl2, errLvl3)) // false
	fmt.Println()
	fmt.Println(errors.Is(errLvl3, errLvl1)) // true
	fmt.Println(errors.Is(errLvl3, errLvl2)) // true
	fmt.Println(errors.Is(errLvl3, errLvl3)) // true
	fmt.Println()
	fmt.Println(errors.Unwrap(errLvl1)) // nil
	fmt.Println(errors.Unwrap(errLvl2)) // error 1
	fmt.Println(errors.Unwrap(errLvl3)) // error 2, because of: error 1
}

```

### PR DESCRIPTION
**Pull request by Psykepro**
Signed-off-by: Nikolay Nedkov <nikolai_nedkov@yahoo.com>

Closes #1059.

### What does this PR do?

To wrap non-wrapped errors in order to be able to check errors properly with `errors.Is()`.

### Reviewers

Main reviewers:

- @tclemos 
- @kind84

Codeowner reviewers:

- @tclemos 
- @ToniRamirezM 
- @Mikelle 
- @KonradIT

### CODE REVIEW

File: state/state_test.go [hunk 1]
```diff
@@ -226,7 +227,7 @@ func TestOpenCloseBatch(t *testing.T) {
 		GlobalExitRoot: common.HexToHash("c"),
 	}
 	err = testState.OpenBatch(ctx, processingCtx3, dbTx)
-	require.True(t, strings.Contains(err.Error(), "unexpected batch"))
+	require.True(t, errors.Is(err, state.ErrUnexpectedBatch))
```
Comment (KonradIT) [inline state/state_test.go:230]:
```suggestion
	require.ErrorIs(t, err, state.ErrUnexpectedBatch)
```

File: test/operations/wait.go [hunk 2]
```diff
@@ -201,11 +201,8 @@ func txMinedCondition(ctx context.Context, client ethClienter, hash common.Hash)
 	}
 	if receipt.Status == types.ReceiptStatusFailed {
 		// Get revert reason
-		reason, reasonErr := revertReason(ctx, client, tx, receipt.BlockNumber)
-		if reasonErr != nil {
-			reason = reasonErr.Error()
-		}
-		return false, fmt.Errorf("transaction has failed, reason: %s, receipt: %+v. tx: %+v, gas: %v", reason, receipt, tx, tx.Gas())
+		_, reasonErr := revertReason(ctx, client, tx, receipt.BlockNumber)
```
Comment (KonradIT) [inline test/operations/wait.go:204]: `reasonErr` could be nil if the revertReason function is correctly executed, and the log says `reason: ` for the reason string.

**Review by KonradIT — CHANGES_REQUESTED** [review]

File: test/operations/wait.go (hunk 2 above)
Comment (Psykepro) [reply test/operations/wait.go:204]: ops yep :D

**Review by KonradIT — APPROVED** [review]
lgtm

**Review by kind84 — APPROVED** [review]
LGTM, nice one @Psykepro :raised_hands:

File: sequencer/sequencesender.go [hunk 3]
```diff
@@ -5,9 +5,9 @@ import (
 	"errors"
 	"fmt"
 	"math/big"
-	"strings"
 	"time"
 
+	ethman "github.com/0xPolygonHermez/zkevm-node/etherman"
```
Comment (ARR552) [inline sequencer/sequencesender.go:10]:
```suggestion
	"github.com/0xPolygonHermez/zkevm-node/etherman"
```

File: sequencer/sequencesender.go [hunk 4]
```diff
@@ -152,7 +152,7 @@ func (s *Sequencer) handleEstimateGasSendSequenceErr(
 	err error,
 ) ([]types.Sequence, error) {
 	// Insufficient allowance
-	if strings.Contains(err.Error(), errInsufficientAllowance) {
+	if errors.Is(err, ethman.ErrInsufficientAllowance) {
```
Comment (ARR552) [inline sequencer/sequencesender.go:155]:
```suggestion
	if errors.Is(err, etherman.ErrInsufficientAllowance) {
```

File: sequencer/sequencesender.go [hunk 5]
```diff
@@ -175,7 +175,7 @@ func (s *Sequencer) handleEstimateGasSendSequenceErr(
 
 	// while estimating gas a new block is not created and the POE SC may return
 	// an error regarding timestamp verification, this must be handled
-	if strings.Contains(err.Error(), errTimestampMustBeInsideRange) {
+	if errors.Is(err, ethman.ErrTimestampMustBeInsideRange) {
```
Comment (ARR552) [inline sequencer/sequencesender.go:178]:
```suggestion
	if errors.Is(err, etherman.ErrTimestampMustBeInsideRange) {
```

File: sequencer/sequencesender.go [hunk 6]
```diff
@@ -215,7 +215,7 @@ func (s *Sequencer) handleEstimateGasSendSequenceErr(
 }
 
 func isDataForEthTxTooBig(err error) bool {
-	return strings.Contains(err.Error(), errGasRequiredExceedsAllowance) ||
+	return errors.Is(err, ethman.ErrGasRequiredExceedsAllowance) ||
```
Comment (ARR552) [inline sequencer/sequencesender.go:218]:
```suggestion
	return errors.Is(err, etherman.ErrGasRequiredExceedsAllowance) ||
```

File: sequencer/sequencesender.go [hunk 7]
```diff
@@ -215,7 +215,7 @@ func (s *Sequencer) handleEstimateGasSendSequenceErr(
...
 		errors.Is(err, core.ErrOversizedData) ||
-		strings.Contains(err.Error(), errContentLengthTooLarge)
+		errors.Is(err, ethman.ErrContentLengthTooLarge)
```
Comment (ARR552) [inline sequencer/sequencesender.go:220]:
```suggestion
		errors.Is(err, etherman.ErrContentLengthTooLarge)
```

File: state/state.go [hunk 8]
```diff
@@ -366,7 +366,7 @@ func (s *State) OpenBatch(ctx context.Context, processingContext ProcessingConte
 		return err
 	}
 	if lastBatchNum+1 != processingContext.BatchNumber {
-		return fmt.Errorf("unexpected batch number %v, should be %v", processingContext.BatchNumber, lastBatchNum+1)
+		return fmt.Errorf("%w number %v, should be %v", ErrUnexpectedBatch, processingContext.BatchNumber, lastBatchNum+1)
```
Comment (ARR552) [inline state/state.go:369]:
```suggestion
		return fmt.Errorf("%w number %d, should be %d", ErrUnexpectedBatch, processingContext.BatchNumber, lastBatchNum+1)
```

File: state/state.go [hunk 9]
```diff
@@ -584,7 +584,7 @@ func (s *State) isBatchClosable(ctx context.Context, receipt ProcessingReceipt,
 		return err
 	}
 	if lastBatchNum != receipt.BatchNumber {
-		return fmt.Errorf("unexpected batch number %v, should be %v", receipt.BatchNumber, lastBatchNum)
+		return fmt.Errorf("%w number %v, should be %v", ErrUnexpectedBatch, receipt.BatchNumber, lastBatchNum)
```
Comment (ARR552) [inline state/state.go:587]:
```suggestion
		return fmt.Errorf("%w number %d, should be %d", ErrUnexpectedBatch, receipt.BatchNumber, lastBatchNum)
```

File: sequencer/batchbuilder.go [hunk 10]
```diff
@@ -271,8 +270,8 @@ func (s *Sequencer) newSequence(ctx context.Context) (types.Sequence, error) {
 	if err != nil {
 		if rollbackErr := dbTx.Rollback(ctx); rollbackErr != nil {
 			return types.Sequence{}, fmt.Errorf(
-				"failed to rollback dbTx when closing batch that gave err: %s. Rollback err: %s",
-				rollbackErr.Error(), err.Error(),
+				"failed to rollback dbTx when closing batch that gave err: %s. Rollback err: %w",
+				rollbackErr.Error(), err,
```
Comment (ARR552) [inline sequencer/batchbuilder.go:274]:
```suggestion
				"failed to rollback dbTx when closing batch that gave err: %w. Rollback err: %w",
				rollbackErr, err,
```

File: ethtxmanager/ethtxmanager.go [hunk 11]
```diff
@@ -6,14 +6,16 @@ package ethtxmanager
...
 	"fmt"
 	"math/big"
-	"strings"
 	"time"
 
 	ethmanTypes "github.com/0xPolygonHermez/zkevm-node/etherman/types"
 	"github.com/0xPolygonHermez/zkevm-node/log"
 	"github.com/0xPolygonHermez/zkevm-node/proverclient/pb"
+	"github.com/0xPolygonHermez/zkevm-node/state/runtime"
+	"github.com/0xPolygonHermez/zkevm-node/test/operations"
```
Comment (ARR552) [inline ethtxmanager/ethtxmanager.go:18]: It is not a good practice import test packages. I know it is imported in the etherman and it's ok for this PR. But we should create an issue to fix it, just to be sure we don't missed it. Can you do it?

File: etherman/etherman.go [hunk 12]
```diff
@@ -273,7 +273,15 @@ func (etherMan *Client) sequenceBatches(opts *bind.TransactOpts, sequences []eth
 
 		batches = append(batches, batch)
 	}
-	return etherMan.PoE.SequenceBatches(opts, batches)
+
+	transaction, err := etherMan.PoE.SequenceBatches(opts, batches)
```
Comment (ARR552) [inline etherman/etherman.go:277]: there are other functions where this parse can be implemented such as the verifyBatch function

**Review by ARR552 — CHANGES_REQUESTED** [review]

File: sequencer/sequencesender.go (hunk 3 above)
Comment (Psykepro) [reply sequencer/sequencesender.go:10]: this is not valid as it collides with the interface inside the package: https://github.com/0xPolygonHermez/zkevm-node/blob/develop/sequencer/interfaces.go#L32

File: sequencer/sequencesender.go (hunk 4 above)
Comment (Psykepro) [reply sequencer/sequencesender.go:155]: this is not valid as it collides with the interface inside the package: https://github.com/0xPolygonHermez/zkevm-node/blob/develop/sequencer/interfaces.go#L32

File: sequencer/sequencesender.go (hunk 5 above)
Comment (Psykepro) [reply sequencer/sequencesender.go:178]: this is not valid as it collides with the interface inside the package: https://github.com/0xPolygonHermez/zkevm-node/blob/develop/sequencer/interfaces.go#L32

File: sequencer/sequencesender.go (hunk 6 above)
Comment (Psykepro) [reply sequencer/sequencesender.go:218]: this is not valid as it collides with the interface inside the package: https://github.com/0xPolygonHermez/zkevm-node/blob/develop/sequencer/interfaces.go#L32

File: sequencer/sequencesender.go (hunk 7 above)
Comment (Psykepro) [reply sequencer/sequencesender.go:220]: this is not valid as it collides with the interface inside the package: https://github.com/0xPolygonHermez/zkevm-node/blob/develop/sequencer/interfaces.go#L32

File: sequencer/batchbuilder.go (hunk 10 above)
Comment (Psykepro) [reply sequencer/batchbuilder.go:274]: You can't wrap 2 errors you can only wrap 1 error. Isn't the `err` the one we use in the check later on?

File: ethtxmanager/ethtxmanager.go (hunk 11 above)
Comment (Psykepro) [reply ethtxmanager/ethtxmanager.go:18]: Totally agree! I was wondering why we have this functionality in the test package when it is used in a lot of other non-test files. But just for this PR will leave it.

File: etherman/etherman.go (hunk 12 above)
Comment (Psykepro) [reply etherman/etherman.go:277]: I've added it in the last commit please check if this is the proper place. :)

Comment (ARR552) [reply etherman/etherman.go:277]: you also need to add it in the Approve

File: etherman/etherman.go [hunk 13]
```diff
@@ -711,9 +711,15 @@ func (etherMan *Client) verifyBatch(opts *bind.TransactOpts, batchNumber uint64,
...
 	)
+	if err != nil {
+		if parsedErr, ok := tryParseContractPoEError(err); ok {
+			err = parsedErr
+		}
+	}
 
 	if err != nil {
 		return nil, err
 	}
```
Comment (ARR552) [inline etherman/etherman.go:722]: you can join both if

**Review by ARR552 — CHANGES_REQUESTED** [review]

File: etherman/etherman.go (hunk 13 above)
Comment (Psykepro) [reply etherman/etherman.go:722]: omg yep my bad

File: etherman/etherman.go [hunk 14]
```diff
@@ -648,8 +656,12 @@ func (etherMan *Client) GetTxReceipt(ctx context.Context, txHash common.Hash) (*
 func (etherMan *Client) ApproveMatic(maticAmount *big.Int, to common.Address) (*types.Transaction, error) {
 	tx, err := etherMan.Matic.Approve(etherMan.auth, etherMan.SCAddresses[0], maticAmount)
 	if err != nil {
+		if parsedErr, ok := tryParseContractPoEError(err); ok {
```
Comment (ARR552) [inline etherman/etherman.go:659]: You are trying to parse erc20 error with Poe function. Maybe you should rename the function with a more generic name. Same comment in other functions
```suggestion
		if parsedErr, ok := tryParseError(err); ok {
```

**Review by ARR552 — CHANGES_REQUESTED** [review]

File: etherman/etherman.go (hunk 14 above)
Comment (Psykepro) [reply etherman/etherman.go]: I've updated it please review it again. :)

**Review by Mikelle — APPROVED** [review]

**Review by tclemos — APPROVED** [review]

File: etherman/errors.go [hunk 15]
```diff
@@ -11,16 +11,25 @@ var (
...
 	//ErrInsufficientAllowance insufficient allowance
 	ErrInsufficientAllowance = errors.New("insufficient allowance")
+	//ErrBothGasPriceAndMaxFeeGasAreSpecified both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified
+	ErrBothGasPriceAndMaxFeeGasAreSpecified = errors.New("both gasPrice and (maxFeePerGas or maxPriorityFeePerGas) specified")
+	//ErrMaxFeeGasAreSpecifiedButLondonNotActive maxFeePerGas or maxPriorityFeePerGas specified but london is not active yet
+	ErrMaxFeeGasAreSpecifiedButLondonNotActive = errors.New("maxFeePerGas or maxPriorityFeePerGas specified but london is not active yet")
+	//ErrNoSigner no signer to authorize the transaction with
+	ErrNoSigner = errors.New("no signer to authorize the transaction with")
 
 	errorsFromPoEContract = map[string]error{
```
Comment (ARR552) [inline etherman/errors.go:21]: Are we also using this `errorsFromPoEContract` for the approval?

**Review by ARR552 — APPROVED** [review]

**Pull request merged**